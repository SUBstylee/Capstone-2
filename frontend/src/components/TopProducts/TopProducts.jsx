import { useEffect } from "react";
import { useDispatch,useSelector } from "react-redux";
import {Link} from 'react-router-dom';
import {Image} from 'react-bootstrap';
import Loader from '../Loader/Loader';
import Rating from '../Rating/Rating';
import Message from '../Message/Message';
import Carousel from 'react-multi-carousel';
import {listTopProducts} from '../../actions/productActions';

import './TopProducts.scss';
import 'react-multi-carousel/lib/styles.css';

const responsive = {
    desktop: {
      breakpoint: { max: 3000, min: 1024 },
      items: 3,
      partialVisibilityGutter: 60,
    },
    tablet: {
      breakpoint: { max: 1024, min: 464 },
      items: 2,
      partialVisibilityGutter: 50,
    },
    mobile: {
      breakpoint: { max: 464, min: 0 },
      items: 1,
      partialVisibilityGutter: 20,
    },
  };

const TopProducts = () => {
    const dispatch=useDispatch();
    const productTopRated=useSelector(state=>state.productTopRated);
    const {loading,error,products} =productTopRated;
    const images = products.map((product) => product);

    useEffect(()=>{
        dispatch(listTopProducts());
    },[dispatch]);

    return (
        <div className="TopProducts">
            {loading?(<Loader/>):error?(<Message variant='danger'>{error}</Message>):(
                <>
                <h3 className='carousel-title m-0 p-3 text-dark'>Top Rated Products</h3>
                <Carousel
                  autoPlay
                  infinite
                  className='image-item fade-in'
                  responsive={responsive}
                  swipeable
                  renderArrowsWhenDisabled
                  partialVisible={true}
                >
                  {images.slice(0, 8).map((product) => {
                    return (
                      <div key={product.name} className=''>
                          <Link to={`/products/${product._id}`} className='text-dark'>
                        <Image
                          fluid
                          className='img-carousel'
                          draggable={false}
                          src={product.image}
                        />
                        
                          <p className='p-0 m-0  fs-6 '>{product.name}</p>
                        </Link>
                        <p className='p-0 m-0  fs-6'>${product.price}</p>
                        <Rating
                          value={product.rating}
                          text={`${product.numReviews}  reviews`}
                        />
                      </div>
                    );
                  })}
                </Carousel>
              </>
            )}
        </div>
    );
};

export default TopProducts