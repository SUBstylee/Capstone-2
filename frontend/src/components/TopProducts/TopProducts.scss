.TopProducts{
    .img-carousel {
        height: 50%;
        border-radius: 5px;
        outline: none;
        box-shadow: none;
        
      }
      
      .image-item {
        padding: 20px 20px 20px 20px;
        background: rgb(255, 255, 255);
        text-align: center;
        outline: none;
        box-shadow: none;
      }
      
      .carousel-title {
        background: rgba(0, 0, 0, 0);
        outline: none;
        box-shadow: none;
        text-align: center;
      }
}