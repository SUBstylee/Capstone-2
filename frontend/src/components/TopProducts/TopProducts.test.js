import { render } from '@testing-library/react';
import TopProducts from './TopProducts';
import store from '../../store';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

test("renders Toast component without crashing", () => {
    const { getByText } = render(
        <Provider store={store}><BrowserRouter><TopProducts test='test' /></BrowserRouter></Provider>
    );

    expect(getByText('test')).toBeInTheDocument();
});

test("matches snapshot", function () {
    const { asFragment } = render(
        <Provider store={store}><BrowserRouter><TopProducts test='test' /></BrowserRouter></Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});