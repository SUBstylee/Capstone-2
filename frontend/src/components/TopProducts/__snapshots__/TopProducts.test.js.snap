// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TopProducts"
  >
    <p>
      test
    </p>
    <div
      class="Loader"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</DocumentFragment>
`;
